base/threadingconstructs.jl

```diff
@@ -175,9 +175,57 @@ end
 function _threadsfor(iter, lbody, schedule)
     lidx = iter.args[1]         # index
     range = iter.args[2]
+    esc_range = esc(range)
+    func = if schedule === :greedy
+        greedy_func(esc_range, lidx, lbody)
+    else
+        default_func(esc_range, lidx, lbody)
+    end
     quote
         local threadsfor_fun
-        let range = $(esc(range))
+        $func
+        if $(schedule === :greedy || schedule === :dynamic || schedule === :default)
+            threading_run(threadsfor_fun, false)
+        elseif ccall(:jl_in_threaded_region, Cint, ()) != 0 # :static
+            error("`@threads :static` cannot be used concurrently or nested")
+        else # :static
+            threading_run(threadsfor_fun, true)
+        end
+        nothing
+    end
+end
+
+function greedy_func(itr, lidx, lbody)
+    quote
+        let r = Ref{Task}(), c = Channel{eltype($itr)}(threadpoolsize(),taskref=r) do ch
+            for item in $itr
+                put!(ch, item)
+            end
+        end
+        function threadsfor_fun(tid)
+            t = r[]
+            while !istaskstarted(t)
+                yield(t)
+            end
+            while isready(c)
+                item = try
+                    take!(c)
+                catch e
+                    # should technically never happen, but could?!
+                    # we don't have a non-blocking `take!`
+                    rethrow(e)
+                end
+                local $(esc(lidx)) = item
+                $(esc(lbody))
+            end
```

Comment:
Because this ultimately relies on another task to produce the next item, it could make sense to explicitly `yield` to that task after the current iteration.

Reply:
With some more thought, the way this is set up now it tries to keep the channel full after a `take!` anyway (the only waiting task to `put!` is notified after all), so this isn't necessary.